src/tensor_ops/cuda_kernels.rs

```diff
@@ -0,0 +1,172 @@
+use crate::{
+    shapes::Shape,
+    tensor::cuda::{Cuda, CudaArray},
+    tensor_ops::ops::{BinaryKernel, UnaryKernel},
+};
+use cudarc::device::{AsKernelParam, CudaSlice, LaunchAsync, LaunchConfig};
+use std::sync::Arc;
+
+pub trait UnaryOpCudaKernel {
+    /// Compiled by build.rs
+    const PTX_SRC: &'static str;
+
+    /// Unique name for the kernel
+    const MODULE_NAME: &'static str;
+
+    /// Name of function in the .cu file
+    const FWD_FN_NAME: &'static str;
+
+    /// Name of function in the .cu file
+    const BWD_FN_NAME: &'static str;
+
+    const ALL_FN_NAMES: [&'static str; 2] = [Self::FWD_FN_NAME, Self::BWD_FN_NAME];
+}
+
+impl<K: UnaryOpCudaKernel + AsKernelParam> UnaryKernel<K, f32> for Cuda {
+    fn forward<S: Shape>(
+        &self,
+        op: K,
+        inp: &Self::Storage<S, f32>,
+    ) -> Result<Self::Storage<S, f32>, Self::Err> {
+        if !self.dev.has_func(K::MODULE_NAME, K::FWD_FN_NAME) {
+            self.dev
+                .load_ptx(K::PTX_SRC.into(), K::MODULE_NAME, &K::ALL_FN_NAMES)?;
+        }
+
+        let numel = inp.data.len();
+        let mut storage = self.dev.alloc_zeros_async::<f32>(numel)?;
+
+        let fwd_fn = self.dev.get_func(K::MODULE_NAME, K::FWD_FN_NAME).unwrap();
+        let cfg = LaunchConfig::for_num_elems(numel as u32);
```

Comment:
Should probably make a helper method for computing a good version of this launch config - need to take advantage of threads & blocks.